test/unit3/test_math.py

```diff
@@ -584,48 +584,51 @@ def testfrexp(name, result, expected):
-        #         if texp > exp:
-        #             tmant <<= texp-exp
-        #             texp = exp
-        #         else:
-        #             mant <<= exp-texp
-        #         tmant += mant
-        #     # Round tmant * 2**texp to a float.  The original recipe
-        #     # used float(str(tmant)) * 2.0**texp for this, but that's
-        #     # a little unsafe because str -> float conversion can't be
-        #     # relied upon to do correct rounding on all platforms.
-        #     tail = max(len(bin(abs(tmant)))-2 - mant_dig, etiny - texp)
-        #     if tail > 0:
-        #         h = 1 << (tail-1)
-        #         tmant = tmant // (2*h) + bool(tmant & h and tmant & 3*h-1)
-        #         texp += tail
-        #     return math.ldexp(tmant, texp)
+    def testFsum(self):
+        # math.fsum relies on exact rounding for correct operation.
+        # There's a known problem with IA32 floating-point that causes
+        # inexact rounding in some situations, and will cause the
+        # math.fsum tests below to fail; see issue #2937.  On non IEEE
+        # 754 platforms, and on IEEE 754 platforms that exhibit the
+        # problem described in issue #2937, we simply skip the whole
+        # test.
+    
+        # Python version of math.fsum, for comparison.  Uses a
+        # different algorithm based on frexp, ldexp and integer
+        # arithmetic.
+        # from sys import float_info
+        # mant_dig = float_info.mant_dig
+        # etiny = float_info.min_exp - mant_dig
+        mant_dig = 53 # sys and float_info don't exist
+        etiny = -1074-mant_dig
+
+        def msum(iterable):
+            """Full precision summation.  Compute sum(iterable) without any
+            intermediate accumulation of error.  Based on the 'lsum' function
+            at http://code.activestate.com/recipes/393090/
+            """
+            tmant, texp = 0, 0
+            for x in iterable:
+                mant, exp = math.frexp(x)
+                mant, exp = int(math.ldexp(mant, mant_dig)), exp - mant_dig
+                if texp > exp:
+                    tmant <<= texp-exp
+                    texp = exp
+                else:
+                    mant <<= exp-texp
+                tmant += mant
+            # Round tmant * 2**texp to a float.  The original recipe
+            # used float(str(tmant)) * 2.0**texp for this, but that's
+            # a little unsafe because str -> float conversion can't be
+            # relied upon to do correct rounding on all platforms.
+            tail = max(len(bin(abs(tmant)))-2 - mant_dig, etiny - texp)
+            if tail > 0:
+                h = 1 << (tail-1)
+                tmant = tmant // (2*h) + bool(tmant & h and tmant & 3*h-1)
+                texp += tail
+            
+            # weird hack that fixes a bug when tmant is converted to a float incorrectly by skulpt
+            return math.ldexp(int(str(tmant)), texp)  
```

Comment:
I wasn't sure why I had to do this but I did?
`tmant` was a large negative `int` and doing `float(tmant)` vs `float(int(str(tmant))` gave wildly different values.
when testing this I had a value `tmat = -2492302492096719`
```
float(tmat)              --> -1689704655.0
float(-2492302492096719) --> -2492302492096719.0
float(str(int(tmat))     --> -2492302492096719.0
```

the adapted code which highlights this issue is... :
```python
import math
from math import fsum

mant_dig = 53
etiny = -1024 - 53


def msum(iterable):
    """Full precision summation.  Compute sum(iterable) without any
    intermediate accumulation of error.  Based on the 'lsum' function
    at http://code.activestate.com/recipes/393090/
    """
    tmant, texp = 0, 0
    for x in iterable:
        mant, exp = math.frexp(x)

        mant, exp = int(math.ldexp(mant, mant_dig)), exp - mant_dig
        if texp > exp:
            tmant <<= texp-exp
            texp = exp
        else:
            mant <<= exp-texp
        tmant += mant
    # Round tmant * 2**texp to a float.  The original recipe
    # used float(str(tmant)) * 2.0**texp for this, but that's
    # a little unsafe because str -> float conversion can't be
    # relied upon to do correct rounding on all platforms.
    tail = max(len(bin(abs(tmant)))-2 - mant_dig, etiny - texp)
    if tail > 0:
        h = 1 << (tail-1)
        tmant = tmant // (2*h) + bool(tmant & h and tmant & 3*h-1)
        texp += tail
    print(tmant, texp)

    print(float(tmant))
    print(float(-2492302492096719))
    print(float(int(str(tmant))))

    return math.ldexp(tmant, texp)


from random import random, gauss, shuffle, seed
seed(0)
# print(random())
for j in range(1):
    vals = [7, 1e100, -7, -1e100, -9e-20, 8e-20] * 3
    # s = 0
    # for i in range(200):
    #     # print(gauss(0, 0.8444218515250481))
    #     # v = gauss(0, 0.8444218515250481) ** 7 - s
    #     v = 0.044072055380373884 ** 2 - s
    #     print(v)
    #     s += v
    #     vals.append(v)
    # shuffle(vals)
    # print(vals)

    s = msum(vals)
```

Reply:
this is really weird.

Reply:
done with commit abc066f 👍